### PR TITLE
Matexpr: DiagonalMatrix and DiagonalOf enhancements

### LINKED ISSUE #12427: DiagonalMatrix assumes shape is square
`DiagonalMatrix` assumes its shape is (row x row):

```
>>> a=DiagonalMatrix(MatrixSymbol('x', 2, 3))
>>> b=DiagonalMatrix(MatrixSymbol('x', 3, 2))
>>> a.shape, b.shape
((2, 2), (3, 3))
```

So it is not surprising that the multiplication may fail;

```
>>> a*b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\matrices\expressions\matexpr.py", line 21, in __sympifyit_wrapper
    b = sympify(b, strict=True)
  File "sympy\core\decorators.py", line 119, in binary_op_wrapper
    return func(self, other)
  File "sympy\matrices\expressions\matexpr.py", line 106, in __mul__
    def __mul__(self, other):
  File "sympy\matrices\expressions\matmul.py", line 38, in __new__
    validate(*matrices)
  File "sympy\matrices\expressions\matmul.py", line 136, in validate
    raise ShapeError("Matrices %s and %s are not aligned"%(A, B))
sympy.matrices.matrices.ShapeError: Matrices DiagonalMatrix(x) and DiagonalMatri
x(x) are not aligned
```

But non-square matrices should be allowed:

```
>>> Matrix(2,3,(1,0,0,0,2,0)) * Matrix(3,2,(4,0,0,5,0,0))
Matrix([
[4,  0],
[0, 10]])
```

The effective matrix size should be the minimum of the given dimensions. When any dimension is symbolic, there should be a way to indicate which one is the smaller. Or if one is symbolic and the other definite then (maybe) the definite one could be taken to be the limiting one. If both are symbolic there should either be an error or else a way to select the smaller of the two dimensions.

A **related issue** is that `DiagonalOf` assumes its shape is (row, 1):

```
>>> DiagonalOf(MatrixSymbol('x', 3, 2)).shape
(3, 1)  <------- that should be (2, 1)
```

### PR DESCRIPTION
**Pull request by smichr**
Fixes #12427 and gives DiagonalOf and DiagonalMatrix docstrings that demonstrate how they behave.

A break from the previous behavior is that concrete indices are assumed to be smaller than symbolic ones and None is returned if the length of the diagonal cannot be determined. (Previously, the number of rows was always taken to be the length of the diagonal and the DiagonalMatrix was assumed to be square.)

Single-index indexing, e.g. `M[3]` is now allowed as long as the number of columns is known. This means that DiagonalOf can always be addressed wth a single index (since its number of columns is always 1).

A couple of inconsequential edits of other files were made as they were noticed along the way.

### CODE REVIEW

**Comment by smichr** [comment]
Should the `size` attribute be called `diagonal_length`?

File: sympy/matrices/expressions/diagonal.py [hunk 1]
```diff
@@ -1,23 +1,147 @@
 from __future__ import print_function, division
 
 from sympy.matrices.expressions import MatrixExpr
-from sympy.core import S, Eq
+from sympy.core import S, Eq, Ge
 from sympy.functions.special.tensor_functions import KroneckerDelta
+
+
 class DiagonalMatrix(MatrixExpr):
+    """DiagonalMatrix(M) will create a create a matrix expression that
```
Comment (szymag) [inline sympy/matrices/expressions/diagonal.py]: I suppose, that something is wrong here: "will create a create a matrix"

File: sympy/matrices/expressions/diagonal.py [hunk 2]
```diff
@@ -1,23 +1,147 @@
...
-        elif eq is S.true:
+        if eq is S.true:
             return self.arg[i, i]
+        elif eq is S.false:
+            return S.Zero
         return self.arg[i, j]*KroneckerDelta(i, j)
 
+
 class DiagonalOf(MatrixExpr):
+    """DiagonalOf(M) will create a create a matrix expression that
```
Comment (szymag) [inline sympy/matrices/expressions/diagonal.py]: Here is the same: "will create a create a matrix"

**Comment by szymag** [comment]
> Should the size attribute be called diagonal_length?

`Size` is not descriptive at all.  In my opinion `diagonal_length` is much better.

File: sympy/matrices/expressions/diagonal.py (hunk 2 above)
Comment (smichr) [inline sympy/matrices/expressions/diagonal.py]: Got both changes and added "Examples" to the docstring and changed the name from `size` to `diagonal_length`.

**Comment by smichr** [comment]
@szymag , thanks for the review. Changes have been made as a new commit.

**Pull request merged**